### PR TITLE
Move merklepatriciatrie related code to mptrie package 

### PR DESCRIPTION
**Pull request by koseoyoung**
splitting PR #2107 (2nd PR) 
Move merklepatriciatrie related code to mptrie package and define multiple interfaces/ modify  abstraction layer

### CODE REVIEW

File: db/trie/trie.go [hunk 1]
```diff
@@ -9,162 +9,62 @@ package trie
 import (
 	"context"
```
[the remark this answers is not in the document]
Comment (koseoyoung) [reply db/trie/trie.go:11]: yes I'm splitting PR #2107 into 4 PRs. 
This 2nd PR is mainly about refactoring. And planning to send out 3rd: introducing cache node, 4th: async mode.

[the remark this answers is not in the document]
Comment (dustinxie) [reply db/trie/trie.go:11]: ok, pls check above 2 comments
i am not sure if those would be potential issue, that is, the trie behavior might change in certain corner case

File: db/trie/mptrie/branchnode.go [hunk 2]
```diff
@@ -75,33 +71,33 @@ func (b *branchNode) children(tr Trie) ([]Node, error) {
...
 		return nil, err
 	}
-	newChild, err := child.delete(tr, key, offset+1)
+	newChild, err := child.delete(key, offset+1)
 	if err != nil {
 		return nil, err
 	}
-	if newChild != nil {
-		return b.updateChild(tr, offsetKey, newChild)
+	if newChild != nil || b.isRoot {
```
[the remark this answers is not in the document]
Comment (CoderZhi) [reply db/trie/mptrie/branchnode.go:85]: no, it won't.

File: db/trie/mptrie/twolayertrie.go [hunk 3]
```diff
@@ -77,18 +82,29 @@ func (tlt *TwoLayerTrie) Stop(ctx context.Context) error {
...
+
+func (tlt *twoLayerTrie) RootHash() []byte {
 	return tlt.layerOne.RootHash()
 }
 
-// SetRootHash sets root hash for layer one trie
-func (tlt *TwoLayerTrie) SetRootHash(rh []byte) error {
-	return tlt.layerOne.SetRootHash(rh)
+func (tlt *twoLayerTrie) SetRootHash(rh []byte) error {
+	for key, lt := range tlt.layerTwo {
```
[the remark this answers is not in the document]
Comment (CoderZhi) [reply db/trie/mptrie/twolayertrie.go:94]: no, it won't.

**Pull request merged**